File: Leetcode/0331 Verify Preorder Serialization of a Binary Tree.ts
function isValidSerialization(preorder: string): boolean {
  let i = 0, nNodesInStack = 0, foundTree = false;
  while (i < preorder.length) {
    if (foundTree) {
      return false;
    }
    let node = '';
    while (i < preorder.length && preorder[i] !== ',') {
      node += preorder[i++];
    }
    i++;
    if (nNodesInStack) {
      nNodesInStack--;
    }
    if (node !== '#') {
      nNodesInStack += 2
    }
    if (nNodesInStack === 0) {
      foundTree = true;
    }
  }
  return nNodesInStack == 0;
}

File: Leetcode/0331 Verify Preorder Serialization of a Binary Tree.py
class Solution:
    def isValidSerialization(self, preorder):
      n_nodes_in_stack, found_tree = 0, False
      for node in preorder.split(','):
        if found_tree: 
          return False
        if n_nodes_in_stack:
          n_nodes_in_stack -= 1
        if node != '#':
          n_nodes_in_stack += 2
        if n_nodes_in_stack == 0:
          found_tree = True
      return n_nodes_in_stack == 0
